bpo-29822: make inspect.isabstract() work during __init_subclass__

Soares opened the pull request:
At the time when an abstract base class' `__init_subclass__` runs, `ABCMeta.__new__` has not yet finished running, so in the presence of `__init_subclass__`, `inspect.isabstract()` can no longer depend only on `TPFLAGS_IS_ABSTRACT`.

Commit:
fix inspect.isabstract() during __init_subclass__

At the time when an abstract base class' __init_subclass__ runs,
ABCMeta.__new__ has not yet finished running, so in the presence of
__init_subclass__, inspect.isabstract() can no longer depend only on
TPFLAGS_IS_ABSTRACT.

Commit:
Added NEWS item

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -285,7 +286,24 @@ def isroutine(object):
 def isabstract(object):
     """Return true if the object is an abstract base class (ABC)."""
-    return bool(isinstance(object, type) and object.__flags__ & TPFLAGS_IS_ABSTRACT)
+    if isinstance(object, type):
```
```python
if isinstance(object, type):
    return False
```
This will decrease an indentation level.

serhiy-storchaka replied:
```python
if not isinstance(object, type):
    return False
```
This will decrease an indentation level.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -285,7 +286,24 @@ def isroutine(object):
+    if isinstance(object, type):
+        if object.__flags__ & TPFLAGS_IS_ABSTRACT:
+            return True
+        elif (issubclass(type(object), abc.ABCMeta)
```
Just `if` rather than `elif`. And you can return `False` if the condition is false for decreasing an indentation level.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -285,7 +286,24 @@ def isroutine(object):
+            # We're likely in the __init_subclass__ of an ABC. ABCMeta.__new__
+            # hasn't finished running yet. We have to search for abstractmethods
+            # by hand. Code copied from ABCMeta.__new__.
+            abstracts = {name
```
It seems to me that collecting names of abstract method is not needed. 
`False` can be returned just after finding any abstract method.

Commit:
integrate review comments

- return earlier to decrease indentation levels
- return immediately upon finding an abstractmethod, instead of
  collecting the whole set

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -285,7 +286,27 @@ def isroutine(object):
+        if getattr(value, "__isabstractmethod__", False):
+            return True
+    for base in object.__bases__:
+        for name in getattr(base, "__abstractmethods__", frozenset()):
```
You can use an empty tuple `()` rather than `frozenset()`. This is shorter and faster.

serhiy-storchaka commented on Lib/test/test_inspect.py:
```diff
@@ -229,6 +229,27 @@ def foo(self):
+            def foo(self):
+                pass
+
+        class ClassExample(AbstractClassExample):
```
The test doesn't cover the case when the abstract method is not in the `__dict__` of final class, but is inherited from base class.

serhiy-storchaka commented on Misc/NEWS:
```diff
@@ -764,6 +764,9 @@ Library
 - Issue #29581: ABCMeta.__new__ now accepts **kwargs, allowing abstract base
   classes to use keyword parameters in __init_subclass__. Patch by Nate Soares.
 
+- Issue #29822: inspect.isabstract() now works during __init_subclass__. Patch
```
Move this at the start of the Library section. Use two spaces between sentences.

serhiy-storchaka commented on Lib/test/test_inspect.py:
```diff
@@ -229,6 +229,27 @@ def foo(self):
 
+    def test_isabstract_during_init_subclass(self):
+        from abc import ABCMeta, abstractmethod
+
```
Use blank lines in test method only to indicate logical sections.

Commit:
Integrated more review comments.

- Added a new test.
- Cleaned up some code style here and there

serhiy-storchaka commented on Lib/test/test_inspect.py:
```diff
@@ -229,6 +229,30 @@ def foo(self):
+        isabstract_checks.clear()
+        class AbstractChild(AbstractClassExample):
+            pass
+        class AbstractGrandchild(AbstractClassExample):
```
What is the difference between `AbstractChild` and `AbstractGrandchild`?

Commit:
fixed typo in the tests

Soares commented:
Oh, my bad, I meant to have AbstractGrandchild inherit AbstractChild. AbstractGrandchild is not strictly necessary, I suppose; I have it in there to make sure that we're walking the whole `__mro__` instead of just the `__bases__`.

agoose77 commented on Lib/inspect.py:
```diff
@@ -285,7 +286,27 @@ def isroutine(object):
+        # TPFLAGS_IS_ABSTRACT should have been accurate.
+        return False
+    # It looks like ABCMeta.__new__ has not finished running yet; we're
+    # probably in __init_subcalss_. We'll look for abstractmethods manually.
```
Pedantic but typo in comment. Should be:

```python
# probably in __init_subclass__. We'll look for abstractmethods manually.
```

Commit:
Typo fix

Commit:
Merge branch 'master' into fix-issue-29822

Soares commented:
Thanks for the review. Fixed the typo, and updated the NEWS file to resolve conflicts.

Soares commented:
Bump. @agoose77, I have fixed the aforementioned typo, are there any outstanding issues as far as you're concerned?

Commit:
Merge branch 'master' into fix-issue-29822

serhiy-storchaka commented on Misc/NEWS:
```diff
@@ -1863,8 +1866,7 @@ Tests
 
-What's New in Python 3.6.0 beta 1?
-==================================
+What's New in Python 3.6.0 beta ==============================
```
Eaten "1?" and a newline.

Commit:
Typo fix

Fixed "1?\n====" that I accidentally killed during the resolution of a merge conflict.

Soares commented:
My bad; thanks for catching that; should be fixed now.

serhiy-storchaka commented on Lib/inspect.py:
```diff
@@ -291,7 +292,27 @@ def isroutine(object):
+        # TPFLAGS_IS_ABSTRACT should have been accurate.
+        return False
+    # It looks like ABCMeta.__new__ has not finished running yet; we're
+    # probably in __init_subcalss__. We'll look for abstractmethods manually.
```
Typo: "`__init_subcalss__`".

Commit:
Typo fix in comment (calss -> class)

Soares commented:
Fixed thanks.

ncoghlan reviewed (approved):
Nice find, and the proposed fix and test case look good to me.

agoose77 commented:
Thanks @Soares LGTM👍

serhiy-storchaka commented:
Thank you for your contribution @Soares. Do you mind to backport the fix to 3.6?

Soares commented:
I'd be happy to backport. Note that https://github.com/python/cpython/pull/527 will also need backporting (in Python 3.6, ABCMeta doesn't work with `__init_subclass__` at all). IIRC, the developer guide says that that PR will need a "needs backport to 3.6" label that only core devs can apply; if that's correct, can you apply that label? Then I'll backport both PRs. Thanks!
